### DISCUSSION TITLE
RUBBER DUCK? UID of Null on Refresh in EventView.js and FriendView.js, useHistory not working (motivating user to refresh-- though, the refresh is bad enough on its own) and new refresh issues in deployed site

### DISCUSSION THREAD

**Discussion by thedigitalmenagerie** [nss-evening-cohort-14/evening-client-side#309] [Capstones!] [answered]
**Honey-Rae Swan**

**Describe your issue:**

Going to another page and back again updates the DOM and does not cause this issue. User is currently motivated to refresh due to DOM not updating, despite using useHistory. In my local repo, when user refreshes, the errors below occur in EventView and FriendView. In my deployed site, the issues below occur in UserView, CategoryView, ItemView, and SingleCategoryView-- due to state being set in App.js. 

I am including these all because I think this is a larger issue with setting state and passing props that is repeated throughout the application. I made the flow chart, including errors, so I could try to the see bigger picture and . I think a rubber duck on my flow chart is probably where I need to start, given the scope. Additionally, I am really lost on state flow through friend views moving forward and don't know if my current flow chart is for the better or worse in regards to this application. 

**Have you looked at other tickets?**
#285 -- this is not happening on any CRUD actions, only when the user happens to refresh in EventsView and FriendView.
#287-- this is why I am thinking I need a rubber duck on my flow chart.
#252-- I am not setting state for this view in App.js, it is being done in the view on Events. 
#302-- I am setting state for categories and items in App.js due to both being needed in CategoriesView and ItemsView due to their relationship. This is not occurring on my local repo but in deployed only, as is mentioned in this ticket.  I did my flow chart and it looks like I am able to move items set state back to itemsView. 

**Error Messages in Local Repo:**
If the user happens to refresh in **EventsView**, I am getting:
TypeError: Cannot read property 'uid' of null
The above error occurred in the <ItemView> component:
Uncaught TypeError: Cannot read property 'uid' of null
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

If the user happens to refresh in **FriendView**, I am getting:
TypeError: Cannot read property 'uid' of null
The above error occurred in the <FriendView> component:
Uncaught TypeError: Cannot read property 'uid' of null
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

**Error Messages in Deployed Site:**
If the user happens to refresh in **CategoriesView**, I am getting:
myCategories:1 GET https://hr-gratitude-capstone.netlify.app/myCategories 404

If the user happens to refresh in **SingleCategoryView**, I am getting:
GET https://hr-gratitude-capstone.netlify.app/myCategories/-MbgPbocx11wCwiHxLso 404

If the user happens to refresh in **ItemsView**, I am getting:
GET https://hr-gratitude-capstone.netlify.app/myItems 404

If the user happens to refresh in **UserView**, I am getting:
GET https://hr-gratitude-capstone.netlify.app/userView 404

**What have you googled?**
(Reviewed the links provided in other issue tickets and some of my own)
https://stackoverflow.com/questions/58038008/how-to-stop-memory-leak-in-useeffect-hook-react
https://reactjs.org/docs/hooks-state.html
https://stackoverflow.com/questions/64159970/why-is-currentuser-uid-undefined
https://stackoverflow.com/questions/29836168/javascript-check-for-multiple-undefined-objects
https://stackoverflow.com/questions/65469282/firebase-data-not-defined-when-used-inside-react-usestate
https://www.pluralsight.com/guides/how-to-unmount-react.js-node
https://dev.to/chilupa/remove-element-from-dom-in-react-way-n2l
https://stackoverflow.com/questions/55324088/why-do-i-have-to-refresh-page-manually-after-delete

**Tried:**
Details of changes made for each error are also linked and in flow chart. 

Trying to prevent friend and userFriend join table from adding twice: 
Adds user as friend with join table but gives error above and user can be added as a friend and all userFriends show in  friendView. If it Try to use logic similar to specific user in App.js, it gives me a 400  error and does not add
In SingleCategoryView:
If I pass it and don't set another state, it gives me a invalid props error for  purchased as string when it expected boolean
TYPE ERROR cannot read categoryName of undefined for every categoryItem card belonging to that category, UNCAUGHT IN PROMISE same as above and setCategories is not a function and does not allow me to update Items with correct Category information associated
passing AND setting another state gives same errors as above
In CategoryView:
If state is here, I cannot share category information with ItemsView.
In ItemsView:
I am currently working on move useState here again.
In EventsView and UserView:
The only way I could get associated cards to read is by passing the user.uid and it is the source of the error. 

**Files:**
https://lucid.app/lucidchart/invitations/accept/inv_b152d880-cc43-404c-85ed-921ac85de07f?viewport_loc=-3358%2C-2034%2C10525%2C5787%2C0_0

https://hr-gratitude-capstone.netlify.app/

https://github.com/thedigitalmenagerie/hr-gratitude-capstone/blob/userFriends/src/Views/PersonalViews/CategoryView.js
https://github.com/thedigitalmenagerie/hr-gratitude-capstone/blob/userFriends/src/Views/PersonalViews/EventView.js
https://github.com/thedigitalmenagerie/hr-gratitude-capstone/blob/userFriends/src/Views/PersonalViews/FriendView.js
https://github.com/thedigitalmenagerie/hr-gratitude-capstone/blob/userFriends/src/Views/PersonalViews/ItemView.js
https://github.com/thedigitalmenagerie/hr-gratitude-capstone/blob/userFriends/src/Views/PersonalViews/SingleCategoryView.js
https://github.com/thedigitalmenagerie/hr-gratitude-capstone/blob/userFriends/src/Views/PersonalViews/UsersView.js

**Comment by drteresavasquez** [comment]
@thedigitalmenagerie will respond with the fix

**Comment by thedigitalmenagerie** [comment, marked as answer]
I have moved items' setState to ItemsView-- this will remove the breaking error in deployed version for this view.

I added an optional chaining operator(?) in EventView, CategoryView, ItemView, FriendView, and UserView:

  useEffect(() => {
    getEvent(user?.uid).then((response) => setEvents(response));
  }, []);

This will allow me to get the user's uid, instead of it being null on first load.

**I am still working toward a solution on the error in deploy for CategoriesView, since it needs to be defined in App.js.**

I am also still working on the leak warning, I am passing props places they shouldn't be going. Following them and cleaning them up.

**Reply by nashvegasnate** [reply]
@thedigitalmenagerie - I was having the exact same issue with "uid of null" on page load and the optional chaining operator came to the rescue again! Thanks for the ticket!
